var/spack/repos/builtin/packages/parallel-netcdf/package.py

```diff
@@ -128,8 +129,49 @@ def autoreconf(self, spec, prefix):
             autoreconf('-iv')
 
     def configure_args(self):
-        args = ['--with-mpi=%s' % self.spec['mpi'].prefix,
-                'SEQ_CC=%s' % spack_cc]
+        # for convenience
+        spec = self.spec
+
+        # mvapich2 will set the MPI env variables,
+        # but that may not be a given. So set them for safety
+        args = ['--with-mpi=%s' % spec['mpi'].prefix,
+                'SEQ_CC=%s' % spack_cc,
+                'MPICC=%s'  % spec['mpi'].mpicc,
+                'MPICXX=%s' % spec['mpi'].mpicxx,
+                'MPIF77=%s' % spec['mpi'].mpifc,
+                'MPIF90=%s' % spec['mpi'].mpifc,
+                ]
+
+        # setup testing (required at configure time)
+        mpiexec = ''
+        mpiexec_flags = ''
+        # this will be an ordered search.
+        # that isn't great - for example, some HPC machines may use
+        # slurm to manage jobs, but another process launcher to run
+        # e.g., Cray can use aprun + slurm, or mpirun + slurm.
+        mpiexecs = ['srun', 'aprun', 'jsrun', 'mpiexec', 'mpirun']
```

Comment:
We could probably turn this into a list of tuples (no need for an additional dictionary then):
```python
mpiexecs = [('srun', '--cpu-bind=cores -c1 -n'),
            ('aprun', '-n'),
            ('jsrun', '-p'),
            ('mpiexec', '-n')]
```
and iterate over it as follows:
```python
for candidate_mpiexec, mpiexec_flags in mpiexecs:
    ...
```

Reply:
It's better to use a `dict` if you expect the launcher to come from a pre-defined multi-valued variant. (which was the original way)

```
variant('mpiexec', values=('srun','aprun', 'jsrun', 'mpiexec'), default='mpiexec')
...
mpiexec_flags={ 'srun' : '-n',
                             'aprun' : -n',
                             'jsrun' : -p',
                             'mpiexec' : '-n' }

mpiexec = spec.varaints['mpiexec']
mpiexec_command = "{0} {1}".format(mpiexec, mpiexec_flags[mpiexec])
```

The gripe is that I am going to have to push stuff into PNetCDF, HDF5 and probably NetCDF to actually produce functional testing. (I do this in HDF5 currently inside a hacked `setup_build_environment`).

Whatever gets decided here will then go into those packages I guess. I'd prefer this stuff get committed to spack/develop though. It's pretty annoying to keep all of these fixes in my fork's branches.

Reply:
Looking in the code, my comment reminded me why I have the list and dict (not just one)
In python3 dicts are ordered, but in 2 they are not.

The list imposes the default precedence, while the dict provides the mapping of command to args.

You could still use a list of tuples and get the ordering... I don't like the syntax or readability though. I should have named my list, `mpiexec_precedences` or something like that, as that is the point.

For now, I've moved to the tuple, but I think ultimately this goes away anyway. The Dicts + Lists are mostly holdovers from wanting the variants that would have made them more logical to use. (e.g., variant value == key to a dict)